apps/heft/src/operations/OperationExecutionManager.ts

```diff
@@ -0,0 +1,235 @@
+   * operations are completed successfully, or rejects when any operation fails.
+   */
+  public async executeAsync(): Promise<void> {
+    const totalOperations: number = this._totalOperations;
+
+    this._terminal.writeLine(`Executing a maximum of ${this._parallelism} simultaneous processes...`);
+
+    const maxParallelism: number = Math.min(totalOperations, this._parallelism);
+    const prioritySort: IOperationSortFunction = (
+      a: OperationExecutionRecord,
+      b: OperationExecutionRecord
+    ): number => {
+      return a.criticalPathLength! - b.criticalPathLength!;
+    };
+    const executionQueue: AsyncOperationQueue = new AsyncOperationQueue(this._executionRecords, prioritySort);
+
+    // This function is a callback because it may write to the collatedWriter before
+    // operation.executeAsync returns (and cleans up the writer)
+    const onOperationComplete: (record: OperationExecutionRecord) => void = (
+      record: OperationExecutionRecord
+    ) => {
+      const endTime: number = performance.now();
+
+      // This operation failed. Mark it as such and all reachable dependents as blocked.
+      if (record.status === OperationStatus.Failure) {
+        // Failed operations get reported, even if silent.
+        // Generally speaking, silent operations shouldn't be able to fail, so this is a safety measure.
+        const message: string | undefined = record.error?.message;
+        if (message) {
+          record.terminal.writeErrorLine(message);
+        }
+        const blockedQueue: Set<OperationExecutionRecord> = new Set(record.consumers);
+        for (const blockedRecord of blockedQueue) {
+          if (blockedRecord.status === OperationStatus.Ready) {
+            blockedRecord.status = OperationStatus.Blocked;
+            for (const dependent of blockedRecord.consumers) {
+              blockedQueue.add(dependent);
+            }
+            if (blockedRecord.runner.groupName) {
+              this._runnerGroupRemainingRecords.get(blockedRecord.runner.groupName)!.delete(blockedRecord);
+            }
+          }
+        }
+        this._hasReportedFailures = true;
+      }
+
+      // At this point, we should only have Success/SuccessWithWarning/Failure cases. Remove
+      // the record and log out the overall time taken for this runner group to complete.
+      const runnerGroupName: string | undefined = record.runner.groupName;
+      if (runnerGroupName) {
+        const groupRemainingRecords: Set<OperationExecutionRecord> =
+          this._runnerGroupRemainingRecords.get(runnerGroupName)!;
+        groupRemainingRecords.delete(record);
+        if (groupRemainingRecords.size === 0) {
+          const startTime: number = this._runnerGroupStartTimes.get(runnerGroupName)!;
+          const executionTime: number = Math.round(endTime - startTime);
+          const hasFailures: boolean = [...this._runnerGroupRecords.get(runnerGroupName)!].some(
```

Comment:
Might be worth extracting the early return helper function that operates on `Iterable`.